### DISCUSSION TITLE
Access to content only for registered users

### DISCUSSION THREAD

**Discussion by malakhovks** [mediacms-io/mediacms#161] [Q&A] [answered]
Is there a way to set some property/settings to ensure that only registered users have access to the content.
So that unregistered users are not be able to watch even public content.

In **settings.py** there is an option for such thing:
`ALLOW_ANONYMOUS_ACTIONS = ["watch"]  # need be a list ["report", "like", "dislike", "watch"]`
Is it possible to do this by changing this parameter for some value?

**Comment by mgogoulos** [comment, marked as answer]
This is not possible out of the box, but it's easy to add through Django's login_required decorator where it forces users to login in order to visit a resource

Edit files/views.py and add  at the beginning of the file


```
from django.contrib.auth.decorators import login_required
```
Then add the @login_required decorator on any view that you want to make accessible only for logged in users. Example, for the media page and the index page

```
@login_required
def view_media(request):
```

```
@login_required
def index(request):
```
and restart mediacms (`systemctl restart mediacms` for local installation, or `docker-compose restart` if you're using Docker)

Now if you are not logged in and visit a media page, it will ask you to login. 

This could be combined with setting

```
ACCOUNT_EMAIL_VERIFICATION = "mandatory"
```
on cms/local_settings.py, so that users need to confirm email upon registering, or even

```
USERS_CAN_SELF_REGISTER = False
```

where users cannot register and you need to add the accounts for them

**Reply by malakhovks** [reply]
Thanks, I'll try to do that!

**Reply by malakhovks** [reply]
It works 👍

**Reply by DecaTec** [reply]
Well, this really works but is a bit complicated to implement.
Any plans for creating anything such a global option which allows to keep the instance private (i.e. not registered users cannot to anything + self registration disabled)?

**Reply by mgogoulos** [reply]
So since there's an interest for this, it could be a nice feature, make a GLOBAL_LOGIN_REQUIRED option available and if set to True require login for all pages. I just experimented without luck with 1-2 existing solutions, will give it another try though

**Comment by mgogoulos** [comment]
This is now part of the core system via https://github.com/mediacms-io/mediacms/pull/204 that was merged
@DecaTec @malakhovks if you want to try it out:

1. git pull # get latest code
2. add GLOBAL_LOGIN_REQUIRED = True
depending on your installation way, for single server install add it on `cms/local_settings.py` and for docker installation on `deploy/docker/local_settings.py`
3. for single server install, run #pip install -r requirements.txt to get the new dependency
4. restart

**Reply by DecaTec** [reply]
Edit: Yes, this works. 
Forgot to recreate the Docker container...


Using Docker:
Pulled latest version, set the variable to `True` and during `docker-compose up`, there's following error:
```
7825938bdafd_mediacms_web_1 | Traceback (most recent call last):
7825938bdafd_mediacms_web_1 |   File "/home/mediacms.io/mediacms/./cms/wsgi.py", line 7, in <module>
7825938bdafd_mediacms_web_1 |     application = get_wsgi_application()
7825938bdafd_mediacms_web_1 |   File "/home/mediacms.io/lib/python3.8/site-packages/django/core/wsgi.py", line 13, in get_wsgi_application
7825938bdafd_mediacms_web_1 |     return WSGIHandler()
7825938bdafd_mediacms_web_1 |   File "/home/mediacms.io/lib/python3.8/site-packages/django/core/handlers/wsgi.py", line 127, in __init__
7825938bdafd_mediacms_web_1 |     self.load_middleware()
7825938bdafd_mediacms_web_1 |   File "/home/mediacms.io/lib/python3.8/site-packages/django/core/handlers/base.py", line 40, in load_middleware
7825938bdafd_mediacms_web_1 |     middleware = import_string(middleware_path)
7825938bdafd_mediacms_web_1 |   File "/home/mediacms.io/lib/python3.8/site-packages/django/utils/module_loading.py", line 17, in import_string
7825938bdafd_mediacms_web_1 |     module = import_module(module_path)
7825938bdafd_mediacms_web_1 |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
7825938bdafd_mediacms_web_1 |     return _bootstrap._gcd_import(name[level:], package, level)
7825938bdafd_mediacms_web_1 | ModuleNotFoundError: No module named 'login_required'
7825938bdafd_mediacms_web_1 | unable to load app 0 (mountpoint='') (callable not found or import error)
7825938bdafd_mediacms_web_1 | *** no app loaded. GAME OVER ***
7825938bdafd_mediacms_web_1 | 2021-06-04 13:08:05,624 INFO exited: uwsgi (exit status 22; not expected)
7825938bdafd_mediacms_web_1 | 2021-06-04 13:08:06,626 INFO spawned: 'uwsgi' with pid 49
7825938bdafd_mediacms_web_1 | [uWSGI] getting INI configuration from /home/mediacms.io/mediacms/deploy/docker/uwsgi.ini
```

So it seems that the module cannot be loaded.
The `pip install...` is only for non-Docker installs, right?

Am I missing something?

**Reply by mgogoulos** [reply]
if you've already using a previous image you have to get the latest (by reading your edit I understand you did this). So 

```
docker pull mediacms/mediacms
```

and then restart docker-compose should be ok!
